File: api/schedule.js
const cheerio = require('cheerio');

const FETCH_HEADERS = {
  'User-Agent': 'Mozilla/5.0 (Linux; Android 10) AppleWebKit/537.36 Chrome/120.0.0.0 Mobile Safari/537.36',
  'Accept-Language': 'ja,en;q=0.5',
};

// ── 日向坂46（HTMLスクレイピング） ──────────────────────────────

function parseHinatazaka(html, year, month, targetDay) {
  const $ = cheerio.load(html);
  const items = [];
  let currentDay = null;

  $('*').each((_, el) => {
    const classes = ($(el).attr('class') || '').split(/\s+/);

    if (classes.includes('c-schedule__date--list')) {
      const day = parseInt($(el).find('span').first().text().trim(), 10);
      if (!isNaN(day)) currentDay = day;
      return;
    }

    if (el.tagName === 'li' && classes.includes('p-schedule__item')) {
      if (currentDay !== targetDay) return;

      const category = $(el).find('.c-schedule__category').text().trim();
      const time     = $(el).find('.c-schedule__time--list').text().trim();
      const title    = $(el).find('.c-schedule__text').text().trim();
      const href     = $(el).find('a').attr('href') || '';

      items.push({
        group: '日向坂46',
        date: `${year}-${String(month).padStart(2, '0')}-${String(targetDay).padStart(2, '0')}`,
        time, category, title,
        url: href ? 'https://www.hinatazaka46.com' + href : '',
        members: [],
      });
    }
  });

  return items;
}

// ── 櫻坂46（HTMLスクレイピング） ────────────────────────────────

function parseSakurazaka(html, year, month, targetDay) {
  const $ = cheerio.load(html);
  const items = [];
  const targetDate = `${year}.${String(month).padStart(2, '0')}.${String(targetDay).padStart(2, '0')}`;

  $('.js-schedule-detail').each((_, el) => {
    // .date の text は "2026.05.04" または "2026.05.0418:30～" の形式
    const rawDate = $(el).find('.date').text().replace(/[\s ]+/g, '');
    if (rawDate.slice(0, 10) !== targetDate) return;

    const time     = rawDate.slice(10).replace(/～$/, '');
    const category = $(el).find('.type').text().trim();
    const title    = $(el).find('h2.title, .title').first().text().trim();
    const href     = $(el).find('.lead a').attr('href') || $(el).find('a[href]').attr('href') || '';
    const url      = href.startsWith('http') ? href : href ? 'https://sakurazaka46.com' + href : '';

    items.push({
      group: '櫻坂46',
      date: `${year}-${String(month).padStart(2, '0')}-${String(targetDay).padStart(2, '0')}`,
      time, category, title, url, members: [],
    });
  });

  return items;
}

// ── 乃木坂46（JSON API） ─────────────────────────────────────────

function parseJsonp(body) {
  const start = body.indexOf('(');
  const end   = body.lastIndexOf(')');
  if (start === -1 || end <= start) return null;
  try {
    return JSON.parse(body.slice(start + 1, end));
  } catch {
    return null;
  }
}

function parseNogizaka(body, year, month, targetDay) {
  const parsed = parseJsonp(body);
  if (!parsed?.data) return [];

  return parsed.data.map(item => {
    // text フィールドは HTML。最初の <a href> からURLを抽出
    const $t = cheerio.load(item.text || '');
    const href = $t('a[href]').first().attr('href') || '';
    const url  = href.startsWith('http') ? href
               : href ? 'https://www.nogizaka46.com' + href
               : '';

    return {
      group: '乃木坂46',
      date: `${year}-${String(month).padStart(2, '0')}-${String(targetDay).padStart(2, '0')}`,
      time:     item.start_time || item.time || '',
      category: item.category_name || item.category || '',
      title:    item.title || '',
      url,
      members:  [],
    };
  });
}

// ── 共通フェッチ ────────────────────────────────────────────────

async function fetchSite(url) {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), 8000);
  try {
    const res = await fetch(url, { headers: FETCH_HEADERS, signal: controller.signal });
    if (!res.ok) return null;
    return await res.text();
  } catch {
    return null;
  } finally {
    clearTimeout(timer);
  }
}

// ── エントリーポイント ───────────────────────────────────────────

module.exports = async (req, res) => {
  const nowJST = new Date(Date.now() + 9 * 60 * 60 * 1000);
  const todayStr = nowJST.toISOString().slice(0, 10).replace(/-/g, '');
  const dateStr = (req.query.date || todayStr).replace(/\D/g, '');

  let year, month, day;
  if (/^\d{8}$/.test(dateStr)) {
    year  = parseInt(dateStr.slice(0, 4), 10);
    month = parseInt(dateStr.slice(4, 6), 10);
    day   = parseInt(dateStr.slice(6, 8), 10);
  } else {
    year  = nowJST.getUTCFullYear();
    month = nowJST.getUTCMonth() + 1;
    day   = nowJST.getUTCDate();
  }

  const [nogiBody, sakuraBody, hinataBody] = await Promise.all([
    fetchSite(`https://www.nogizaka46.com/s/n46/api/list/schedule?dy=${dateStr}`),
    fetchSite(`https://sakurazaka46.com/s/s46/media/list?dy=${dateStr}`),
    fetchSite(`https://www.hinatazaka46.com/s/official/media/list?dy=${dateStr}`),
  ]);

  const items = [
    ...(nogiBody   ? parseNogizaka(nogiBody,   year, month, day) : []),
    ...(sakuraBody ? parseSakurazaka(sakuraBody, year, month, day) : []),
    ...(hinataBody ? parseHinatazaka(hinataBody, year, month, day) : []),
  ];

  items.sort((a, b) => {
    const ta = a.time || '99:99';
    const tb = b.time || '99:99';
    return ta < tb ? -1 : ta > tb ? 1 : a.group.localeCompare(b.group);
  });

  res.setHeader('Content-Type', 'application/json; charset=utf-8');
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.status(200).json(items);
};

File: api/schedule.py
from http.server import BaseHTTPRequestHandler
from bs4 import BeautifulSoup, Tag
import json
import requests
import urllib.parse
import concurrent.futures
from datetime import datetime, timezone, timedelta

HEADERS = {
    'User-Agent': 'Mozilla/5.0 (Linux; Android 10) AppleWebKit/537.36 Chrome/120.0.0.0 Mobile Safari/537.36',
    'Accept-Language': 'ja,en;q=0.5',
}

SITES = [
    {
        'group': '乃木坂46',
        'url': 'https://www.nogizaka46.com/s/n46/media/list?dy={date}',
        'base_url': 'https://www.nogizaka46.com',
    },
    {
        'group': '櫻坂46',
        'url': 'https://sakurazaka46.com/s/s46/media/list?dy={date}',
        'base_url': 'https://sakurazaka46.com',
    },
    {
        'group': '日向坂46',
        'url': 'https://www.hinatazaka46.com/s/official/media/list?dy={date}',
        'base_url': 'https://www.hinatazaka46.com',
    },
]


def parse_schedule(html, group, base_url, year, month, target_day):
    soup = BeautifulSoup(html, 'html.parser')
    items = []
    current_day = None

    for element in soup.find_all(True):
        if not isinstance(element, Tag):
            continue
        classes = element.get('class', [])

        if 'c-schedule__date--list' in classes:
            span = element.find('span')
            if span:
                try:
                    current_day = int(span.get_text(strip=True))
                except ValueError:
                    pass

        elif element.name == 'li' and 'p-schedule__item' in classes:
            if current_day != target_day:
                continue

            cat_el = element.find(class_='c-schedule__category')
            time_el = element.find(class_='c-schedule__time--list')
            title_el = element.find(class_='c-schedule__text')
            link_el = element.find('a', href=True)

            items.append({
                'group': group,
                'date': f'{year:04d}-{month:02d}-{target_day:02d}',
                'time': time_el.get_text(strip=True) if time_el else '',
                'category': cat_el.get_text(strip=True) if cat_el else '',
                'title': title_el.get_text(strip=True) if title_el else '',
                'url': base_url + link_el['href'] if link_el else '',
                'members': [],
            })

    return items


def fetch_site(site, date_str, year, month, target_day):
    try:
        url = site['url'].format(date=date_str)
        r = requests.get(url, headers=HEADERS, timeout=8)
        r.raise_for_status()
        r.encoding = 'utf-8'
        return parse_schedule(r.text, site['group'], site['base_url'], year, month, target_day)
    except Exception:
        return []


class handler(BaseHTTPRequestHandler):
    def do_GET(self):
        qs = urllib.parse.parse_qs(urllib.parse.urlparse(self.path).query)
        jst = timezone(timedelta(hours=9))
        today = datetime.now(jst)
        date_str = qs.get('date', [today.strftime('%Y%m%d')])[0]

        try:
            dt = datetime.strptime(date_str, '%Y%m%d')
        except ValueError:
            dt = today

        y, m, d = dt.year, dt.month, dt.day

        results = []
        with concurrent.futures.ThreadPoolExecutor(max_workers=3) as ex:
            futs = [ex.submit(fetch_site, s, date_str, y, m, d) for s in SITES]
            for f in concurrent.futures.as_completed(futs):
                results.extend(f.result())

        results.sort(key=lambda x: (x['time'] or '99:99', x['group']))

        body = json.dumps(results, ensure_ascii=False).encode('utf-8')
        self.send_response(200)
        self.send_header('Content-Type', 'application/json; charset=utf-8')
        self.send_header('Access-Control-Allow-Origin', '*')
        self.end_headers()
        self.wfile.write(body)

    def log_message(self, *args):
        pass
